[v0.37] Feat / LM strategy updates

Nullably opened the pull request:
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
* Update allocation (see Allocation Calculation in https://www.notion.so/hummingbot/Liquidity-Mining-strategy-0c6eb43ac5ba4c4685e3092f404e0492)
* Add `inventory_skew_enabled` config parameter, this is on by default, but now user can turn it off

Nullably commented:
https://github.com/CoinAlpha/hummingbot/pull/2959 needs to be merged before this PR

mifeng commented:
Requested changes:
1. **Budget allocation**: In `status`, extend the Markets to shows how the user's assets are allocated toward to each market. I think these additional columns should fit in the existing table, but feel free to create a new table.

| Market | Budget (USDT) | Base % | Quote % | 
|--------|-------------------|---------|----------|
|AVAX-USDT| 1000.00 | 0% | 100% |
|XEM-USDT| 1000.00 | 0% | 95% |

2. **Miner status**: In `status`, add another table below it that shows the info from the Miner API. For now, I think we should only display the current status and not automate any bot decisions based on this data.

*As of [time in UTC]:*
| Market | Reward/day | Liquidity (bots) | Yield/day | Max spread|
|--------|-------------|----------------|-----------|------------|
|AVAX-USDT| 1.71 AVAX | $25,395 (19) | 0.25% | 2.00%|
|XEM-USDT| 4600 XEM | $33,008 (19) | 0.79% | 1.00%|

If some of these endpoints aren't available yet, just show what we have currently.

Nullably commented:
I updated the branch to account for the requested changes.

mifeng commented:
<img width="483" alt="Screen Shot 2021-02-22 at 8 47 09 AM" src="https://user-images.githubusercontent.com/1795030/108740532-9f607480-74ea-11eb-8848-32351460e16b.png">

Some API endpoints don't seem to be working, and there's a duplicate % in spread

mifeng commented:
@Nullably I created a sheet creates a revised version of `status` command that attempts more clearly lay out the relevant metrics in a more digestable way: https://docs.google.com/spreadsheets/d/1lYZt85Nhc0hahVyQenFvpguEzb4ozk1hh-A4XYVJB4k/edit?usp=sharing

Nullably commented:
> <img alt="Screen Shot 2021-02-22 at 8 47 09 AM" width="483" src="https://user-images.githubusercontent.com/1795030/108740532-9f607480-74ea-11eb-8848-32351460e16b.png">
> 
> Some API endpoints don't seem to be working, and there's a duplicate % in spread

This data is from development miner API, not real production data. I think Aram is deploying to production soon. I'll update the strategy later.

Nullably commented:
> @Nullably I created a sheet with a revised version of `status` command that attempts more clearly lay out the relevant metrics in a more digestable way: https://docs.google.com/spreadsheets/d/1lYZt85Nhc0hahVyQenFvpguEzb4ozk1hh-A4XYVJB4k/edit?usp=sharing

@mifeng I am not sure what are Best Bid % and Best Ask %, are they distance from mid price (in %)? If so, they will always be the same %.
![image](https://user-images.githubusercontent.com/37262506/108806578-0e1dfa80-75dd-11eb-8a3b-4fbdf0b61f17.png)

PtrckM commented:
- `max_order_age` is not triggering
![image](https://user-images.githubusercontent.com/54516858/108838688-efd0f280-760e-11eb-804e-81f1184d42e4.png)
```
2021-02-23 10:52:46,002 - 3584 - hummingbot.strategy.liquidity_mining.liquidity_mining - INFO - (LUNA-USDT) Creating a bid order [ p: 5.84170000 s: 2.24000000 ] value: 13.09 USDT spread: 5$
2021-02-23 10:52:46,147 - 3584 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Created LIMIT_MAKER BUY order x-XEKWYICX-BLAUT1614077566002389 for 2.24000000 LUNA-USDT.
2021-02-23 10:52:46,157 - 3584 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1614077566.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "LUNA-USDT", "amount": "2.2$
2021-02-23 10:57:47,001 - 3584 - hummingbot.strategy.liquidity_mining.liquidity_mining - INFO - LUNA-USDT volatility: 2.25%
2021-02-23 11:01:06,002 - 3584 - hummingbot.strategy.liquidity_mining.liquidity_mining - INFO - (LUNA-USDT) Cancelling the limit order x-XEKWYICX-BLAUT1614077566002389. [clock=2021-02-23 1$
2021-02-23 11:01:06,149 - 3584 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Successfully cancelled order x-XEKWYICX-BLAUT1614077566002389.
2021-02-23 11:01:06,158 - 3584 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1614078066.0, "order_id": "x-XEKWYICX-BLAUT1614077566002389", "exchange_order_id": null, "$
2021-02-23 11:01:07,001 - 3584 - hummingbot.strategy.liquidity_mining.liquidity_mining - INFO - (LUNA-USDT) Creating a bid order [ p: 5.96020000 s: 2.17000000 ] value: 12.93 USDT spread: 5$
2021-02-23 11:01:07,151 - 3584 - hummingbot.connector.exchange.binance.binance_exchange - INFO - Created LIMIT_MAKER BUY order x-XEKWYICX-BLAUT1614078067002317 for 2.17000000 LUNA-USDT.
2021-02-23 11:01:07,159 - 3584 - hummingbot.core.event.event_reporter - EVENT_LOG - {"timestamp": 1614078067.0, "type": "OrderType.LIMIT_MAKER", "trading_pair": "LUNA-USDT", "amount": "2.1$
2021-02-23 11:02:48,001 - 3584 - hummingbot.strategy.liquidity_mining.liquidity_mining - INFO - LUNA-USDT volatility: 2.30%
```
- `status --live` fails to function properly

https://user-images.githubusercontent.com/54516858/108838881-3a526f00-760f-11eb-9af1-24f028305a97.mp4

Nullably commented:
@PtrckM 
I have updated the branch and here are rules on order canceling in priority order.
1. If order is older than max_order_age, they will get canceled.
2. If order is older than order_refresh_time and spread is not within tolerance, they will be canceled.

I also sorted values on `status` by `Market`.

Nullably commented:
`status --live` will not work with this strategy, the reason being the status command makes a number of API request for Minor information which takes longer time than 1 second. So I think we should mention to user to not use this command with --live.

PtrckM commented:
- hitting status multiple times bot will throw this error
```
2021-02-24 06:56:54,161 - 7353 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task: string indices must be integers
Traceback (most recent call last):
  File "/home/ubuntu/lm_strategy/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/ubuntu/lm_strategy/hummingbot/client/command/status_command.py", line 135, in status_check_all
    self._notify(await self.strategy_status())
  File "/home/ubuntu/lm_strategy/hummingbot/client/command/status_command.py", line 77, in strategy_status
    st_status = await self.strategy.format_status()
  File "/home/ubuntu/lm_strategy/hummingbot/strategy/liquidity_mining/liquidity_mining.py", line 225, in format_status
    miner_df = await self.miner_status_df()
  File "/home/ubuntu/lm_strategy/hummingbot/strategy/liquidity_mining/liquidity_mining.py", line 196, in miner_status_df
    campaigns = await get_campaign_summary(self._exchange.display_name, list(self._market_infos.keys()))
  File "/home/ubuntu/lm_strategy/hummingbot/connector/parrot.py", line 27, in get_campaign_summary
    campaigns = await get_active_campaigns(exchange, trading_pairs)
  File "/home/ubuntu/lm_strategy/hummingbot/connector/parrot.py", line 64, in get_active_campaigns
    for market in campaign_retval["markets"]:
TypeError: string indices must be integers
```
![image](https://user-images.githubusercontent.com/54516858/108964579-9a502080-76b6-11eb-8f3e-461686139d73.png)

PtrckM commented:
**Suggestion:**
- when hitting `max_order_age`
```
06:40:45 - liquidity_mining - (AVAX-USDT) Cancelling the limit order x-XEKWYICX-BAXUT1614148825005114. [clock=2021-02-24 06:40:45+00:00]
06:40:45 - liquidity_mining - (AVAX-USDT) Cancelling the limit order x-XEKWYICX-SAXUT1614148825005304. [clock=2021-02-24 06:40:45+00:00]
06:40:45 - binance_exchange - Successfully cancelled order x-XEKWYICX-BAXUT1614148825005114.
06:40:45 - binance_exchange - Successfully cancelled order x-XEKWYICX-SAXUT1614148825005304.
```
- suggest to have similar cancellation message 
```
14:48:52 - pure_market_making - Refreshing Buy order with ID - x-XEKWYICX-BLAUT1614149311995261 because it reached maximum order age of 20.0 seconds.
14:48:52 - pure_market_making - (LUNA-USDT) Cancelling the limit order x-XEKWYICX-BLAUT1614149311995261. [clock=2021-02-24 06:48:52+00:00]
14:48:52 - pure_market_making - Refreshing Sell order with ID - x-XEKWYICX-SLAUT1614149311996263 because it reached maximum order age of 20.0 seconds.
14:48:52 - pure_market_making - (LUNA-USDT) Cancelling the limit order x-XEKWYICX-SLAUT1614149311996263. [clock=2021-02-24 06:48:52+00:00]
```

mifeng reviewed (dismissed):
I added a commit to revise the labels in `status` output

LGTM

mifeng commented:
> **Suggestion:**
> 
> * when hitting `max_order_age`
> 
> ```
> 06:40:45 - liquidity_mining - (AVAX-USDT) Cancelling the limit order x-XEKWYICX-BAXUT1614148825005114. [clock=2021-02-24 06:40:45+00:00]
> 06:40:45 - liquidity_mining - (AVAX-USDT) Cancelling the limit order x-XEKWYICX-SAXUT1614148825005304. [clock=2021-02-24 06:40:45+00:00]
> 06:40:45 - binance_exchange - Successfully cancelled order x-XEKWYICX-BAXUT1614148825005114.
> 06:40:45 - binance_exchange - Successfully cancelled order x-XEKWYICX-SAXUT1614148825005304.
> ```
> 
> * suggest to have similar cancellation message
> 
> ```
> 14:48:52 - pure_market_making - Refreshing Buy order with ID - x-XEKWYICX-BLAUT1614149311995261 because it reached maximum order age of 20.0 seconds.
> 14:48:52 - pure_market_making - (LUNA-USDT) Cancelling the limit order x-XEKWYICX-BLAUT1614149311995261. [clock=2021-02-24 06:48:52+00:00]
> 14:48:52 - pure_market_making - Refreshing Sell order with ID - x-XEKWYICX-SLAUT1614149311996263 because it reached maximum order age of 20.0 seconds.
> 14:48:52 - pure_market_making - (LUNA-USDT) Cancelling the limit order x-XEKWYICX-SLAUT1614149311996263. [clock=2021-02-24 06:48:52+00:00]
> ```

I actually like the way that @Nullably did log messages in `liquidity_mining` since that is easier to generalize across strategies and connectors. It would be good to log more details from parameters used, but I think we should first try to ensure standardization across the codebase.

mifeng commented:
@Nullably We discussed the Miner API issue during Miner standup, and Aram will try to deploy what's on staging to prod over the weekend so that QA can start testing on Monday.

Nullably commented:
@PtrckM @mifeng 
I have updated the PR to reflect on recent miner public API, the APY is still slightly different to what you get from miner app but I think Aram will work on that soon. 
I also updated in case there is an error from API, we just log the error and the whole Miner section on status will not be visible for now.

CrimsonJacket commented on hummingbot/connector/parrot.py:
```diff
@@ -0,0 +1,94 @@
+class CampaignSummary:
+    market_id: int = 0
+    trading_pair: str = ""
+    exchange_name: str = 0
```
If `exchange name` is a `str`, so the default value shouldn't be an `int`

CrimsonJacket commented on hummingbot/connector/parrot.py:
```diff
@@ -0,0 +1,94 @@
+
+async def get_market_snapshots(market_id: int):
+    async with aiohttp.ClientSession() as client:
+        url = f"{PARROT_MINER_BASE_URL}market_snapshots/{market_id}?aggregate=1m"
```
I was thinking if we should the process of how we create new exchange connectors; with their own `*_constants.py` for stuff like URLs.

I suppose this is special, considering that `parrot.py` functions more like a `util.py`

Nullably replied:
Yeah, it is still small for now, we could refactor this into its own folder and more specific files later once we know more on what we need from this connector.

CrimsonJacket commented on hummingbot/strategy/liquidity_mining/liquidity_mining.py:
```diff
@@ -100,59 +108,105 @@ def tick(self, timestamp: float):
+        df.sort_values(by=["Market"], inplace=True)
+        return df
+
+    async def miner_status_df(self) -> pd.DataFrame:
```
Small matter. I think a better name for `miner_status_df` is `campaign_status_df`.

Nullably replied:
This is to keep it inline with `status` as it is under `Miner` section.

CrimsonJacket reviewed (approved):
Apart from some comments, this PR looks good to me. Minor changes requested for `CampaignSummary` but I will approve it nonetheless.

PtrckM commented:
**Tested and confirmed working by QA.**

- when hitting `status` command multiple times, if parrot api is not yet available bot will not show the miner section on status and throw this message on the right pane.
![image](https://user-images.githubusercontent.com/54516858/109623086-c1579800-7b77-11eb-80bd-882ca87e94cc.png)
- when its ready
![image](https://user-images.githubusercontent.com/54516858/109623537-35923b80-7b78-11eb-819d-6a918d8007d4.png)
- order creation and order cancel have the same messaging when hitting max order age.
- the miner section slightly different from the miner app (expected since still on WIP)

rxlxrxsx reviewed (approved):
Confirmed working and no other issues found by QA.
